Add metrics from SNTPv4 packet to ntp collector & add ntpd sanity check

darkk opened the pull request:
There was no much activity in the mailing list regarding [the future of ntp collector](https://groups.google.com/forum/#!topic/prometheus-users/cGj_BS1KSVM). So I suggest this PR for review. IMHO, beevik/ntp#15 should converge before considering the merge.

1. Checking local clock against remote NTP daemon is bad idea, local ntpd acting as a client should do it better and avoid excessive load on remote NTP server so the collector is refactored to query local NTP server.

2. Checking local clock against remote one does not check local ntpd itself. Local ntpd may be down or out of sync due to network issues, but clock will be OK.

3. Checking NTP server sanity using it's response is tricky and depends on ntpd implementation, that's why common `node_ntp_sanity` variable is exported as an addition to various state variables. 

`node_ntp_drift_seconds` is removed as it can't be reliably guessed from local ntpd: some ntpd implementations expose ntp-clock in T2 & T3 (chrony), some of them expose local clock and indicate local clock drift using LeapNotInSync and stratum (openntpd), some of them use root_dispersion as an indicator of bad local clock (ntp) and some do not (openntpd does not use it at all). I consider it's safe to drop this metrics as the collector was suggested for deprecation and, probably, removal before this refactoring.

This branch was tested against openntpd=20080406p-10, ntp=4.2.6.p5+dfsg-7+deb8u2 and chrony=1.30-2+deb8u2 from Debian jessie and proved to be useful against several failures (unfortunate `date --set`, loss of connectivity).

SuperQ commented:
The original module was intended to be an independent NTP client check, hence then name `ntp`.  I would suggest renaming the entire collector to `ntpd`, to reflect the change in the change in functionality.

matthiasr commented:
From reading the flag descriptions, it's not immediately clear to me which situations are supported and which combination of flags I need to set for them. I agree that if there is a local NTP daemon and we can find out whether things are sane from it, we should.

I'm not sure if the current method (compare local time to NTP) with the new method (check NTP within itself) should even be the same collector. I think it might be easier to understand if there was an `ntp` _client_ and `ntpd` _server_ collector, with an explanation in the README for each of them (and when to use which).

matthiasr commented:
Which NTP daemons can be used with this? Do they need any non-default configuration? This also needs to be documented.

darkk commented:
> which situations are supported

Default is to monitor ntpd running at `127.0.0.1`. You may also want to monitor ntpd running at the same host, but having some different IP (e.g. you run ntpd in docker container for some reason). You have to de-fuse first safeguard and certify that the address you specify is local (set `collector.ntp.server-is-local`). If you have complex network topology at localhost, you have to de-fuse the second safeguard and raise `collector.ntp.ip-ttl` from default value, TTL=1 allows to send packets only to directly-reachable hosts (without intermediate IP router).

> compare local time to NTP

**IF** we're going to keep `ntp` collector that compares local clock to remote NTP server and add this SNTP logic as separate `ntpd` collector, should these two safeguards (that are there to protect NTP pool from being polled every ${scrape_interval}) be also extracted to separate module and implemented in `ntp` collector?

> Which NTP daemons

I've tested ntp.org, openntpd and chrony. I don't know any other wide-spread NTP daemons for Linux.

> Do they need any non-default configuration

Yep, openntpd does not listen by default at all, chrony allows SNTP queries from RFC1918 subnets, but does not allow 127.0.0.0/8. Is `README.md` right place for documentation?

discordianfish commented:
The ntp collector is deprecated. Sorry for not responding to your email to clarify that.
But I think this can be turned into a standalone exporter which we can link to in the README for people looking for replacement for the ntp functionality. Would that work for you? See https://github.com/prometheus/node_exporter/pull/321 for the discussion around this.

darkk commented:
> See #321 

Yep, I've read it and mentioned that fact in the email. This patch addresses following issues:

> The NTP collector is even proxying metrics from a third party system
> Best would be to actually monitor your NTP client
> Some small part of me wants to keep the ntp function in node_exporter. It's just a strange duck.

So this patch refactors NTP collector to check sanity of local NTP client.

It is done to use SNTP query to support three major ntp client implementations (and, probably, other), that's also the reason for `node_ntp_sanity` to "converge" slowly in some cases, e.g. in case of connectivity loss simulated by `iptables ! --in-interface lo -p udp --sport 123 -j DROP`
- it takes chrony 2.5h to indicate that it is out of sync via leap=3, stratum=15, root_delay=1.0, root_dispersion=1.0, refid=`\0\0\0\0`, reftime=0.0
- ntp.org indicates that with root_dispersion linear increase at ~0.054 s/hour (that probably depends on RTT to root), so it takes ~62 hours to hit default max-distance limit and set `sanity` to 0, it also does not update announced reftime so freshness will raise alert in (1 << maxPoll) seconds that is ~36h
- openntpd does not account root_dispersion, but it keeps sane reftime, so it also looses sanity in 36h

I don't consider this collector as a part of blackbox_exporter as blackbox exporter is usually deployed on separate box and:
- this collector needs access **both** to kernel clock and ntp-clock to check them against each other (look at `t21` and `t43`), it works with chrony and openntpd ([served time is not fixed to local time](https://chrony.tuxfamily.org/comparison.html))
- it is considered a good practice to avoid exposing node:123 for bare clients, so checking local NTP client differs from checking other "common" network services like ssh that should be reachable from outside
- (suggested in #321) comparison of ntp timestamp to node_time sounds like **wrong** idea to me: as far as I understand ntpd from ntp.org always shows kernel time as it's served ntp time, but it indicates sync failures using other means: leap flag, stratum, root_dispersion, reftime. Also, you have to query node_time twice to get bounds for packet that will replace T1 and T4 from local SNTP query.

@discordianfish do you still think that it should be separate collector used just for checking local NTP daemon sanity? As far as I understand, this PR addresses actually **the main reason** for the ntp collector being out of scope of node_exporter, it does not rely on remote NTP server anymore.

darkk commented:
Moving sanity checking logic to beevik/ntp as suggested by @discordianfish in IRC.

discordianfish commented:
Just to provide context for people not reading IRC: We decided to accept this functionality but prefer having the checking logic outside of the node-exporter.

darkk commented:
[On the other hand](https://gist.github.com/darkk/1b8c2b188ec2937d3045afe8fc7f645d)...
- chrony 1.30 clears STA_UNSYNC flag if and only if `rtcsync` is enabled, moreover, it clears the flag during SYS_Linux_Initialise, so, seems, it does not reflect synchronization state
- neither openntpd-20080406p nor openntpd-5.7p4 use STA_UNSYNC at all, 5.9p1 is the first portable version using this flag

These versions were checked as they were default packaged versions of corresponding daemons in Debian 8.x (jessie). That's `oldstable` now, but it's still the default Debian version in my environment :-/

- systemd-timesyncd 229-4ubuntu19 clears STA_UNSYNC even if offset is quite high (e.g. 1.4s)
- Linux 4.12-rc1 reads STA_UNSYNC *only for purposes* to return TIME_ERROR from `adjtimex()` and to run `sync_cmos_clock()` if the flag is cleared. Also kernel sets STA_UNSYNC on various events including `settimeofday()`, various errors and `maxerror` overflowing 16s. `maxerror` is accumulator that gets 0.5ms every second in `second_overflow()`

So I consider that both `timex` and this collector are useful given variety of NTP daemons implementations and their interaction with STA_UNSYNC flag.

discordianfish commented:
Okay, I guess this makes sense.

darkk commented:
> prefer having the checking logic outside of the node-exporter

OK, that has finally converged.

I would like to know...
- is it ok to reuse `ntp` namespace for this collector or should I move it to `ntpd`?
- is README the right place to document the collector?
- should it report `sanity` equal to zero when ntpd is down and Query returns error?
- should Query error be logged? Stopped ntpd causes node_exporter to write a line to log every ${scrape_interval} if the collector is enabled.

SuperQ commented:
Query errors should be sent to Debug logs.

darkk commented:
@SuperQ I mean that current code returns `error` from `Update()` (like previous version did) and it leads to following logs while ntpd is down:
```
Sep 01 12:20:11 chrony.tmp.ooni.io node_exporter[5531]: time="2017-09-01T12:20:11Z" level=error msg="ERROR: ntp collector failed after 0.000230s: couldn't get SNTP reply: read udp 127.0.0.1:37385->127.0.0.1:123: read: connection refused" source="node_exporter.go:94"
Sep 01 12:20:16 chrony.tmp.ooni.io node_exporter[5531]: time="2017-09-01T12:20:16Z" level=error msg="ERROR: ntp collector failed after 0.000417s: couldn't get SNTP reply: read udp 127.0.0.1:44391->127.0.0.1:123: read: connection refused" source="node_exporter.go:94"
Sep 01 12:20:21 chrony.tmp.ooni.io node_exporter[5531]: time="2017-09-01T12:20:21Z" level=error msg="ERROR: ntp collector failed after 0.000304s: couldn't get SNTP reply: read udp 127.0.0.1:56474->127.0.0.1:123: read: connection refused" source="node_exporter.go:94"
Sep 01 12:20:26 chrony.tmp.ooni.io node_exporter[5531]: time="2017-09-01T12:20:26Z" level=error msg="ERROR: ntp collector failed after 0.000227s: couldn't get SNTP reply: read udp 127.0.0.1:45065->127.0.0.1:123: read: connection refused" source="node_exporter.go:94"
```

Seems, severity of this `error` is Error, not Debug. My question is -- is it good and expected behavior for node_exporter collectors?

discordianfish commented:
- looking good in general, just a few comments.

Regarding your questions:
- ntp namespace is fine

- yes, README is the place for docs right now but it's in general a bit sparse, so if you like to document more details than 1-2 sentences in the table it's probably best to create docs/NTP.md and link to it in the table (open to other suggestions)

- logging with error is fine I'd say: the exporter isn't enabled by default, so if someone enables it and it fails it's an actionable error. I'd also not report sanity on query errors. You can detect the failing collector 
 via `node_scrape_collector_success` metric. This way you even know whether the collector failed (because query) or your ntp is known to be "insane"

darkk commented:
Thanks. Now I'm waiting for some criticism regarding the PR (both logic and code) as it's the first Go code chunk I've ever written :)

discordianfish commented:
@darkk The metric names still need to be changed, see my comment above. Will review once that's done.

darkk commented:
@discordianfish I'm sorry, what comment are you referring to? I don't see alike comments from you in this PR... Can you cite it, please?

discordianfish commented:
Sorry, got confused by the github review tool. Looks like I didn't submit them. Here they are!

darkk commented on collector/ntp.go:
```diff
@@ -17,62 +17,146 @@ package collector
+			nil, nil,
+		), prometheus.GaugeValue},
+		leap: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "ntp", "leap"),
```
No, it's [0, 1, 2, 3](https://github.com/beevik/ntp/blob/74e5133786235ede462c83c17fa3b34653901719/ntp.go#L38-L50).
I decided to export raw value instead of mapping to something like `0, +1, -1, NaN`.

discordianfish replied:
Ok in that case I'd keep the raw values but still would rename it to ntp_leap_flag

brian-brazil replied:
There is no suffix expected on gauges that happen to be booleans. Unless this is called flag in the code (which it isn't afaics), there shouldn't be a _flag suffix here.

https://godoc.org/github.com/beevik/ntp#LeapIndicator

darkk replied:
BTW, RFC names this field `leap` or `leap indicator`. I think that `flag` may be confusing as flag is usually binary, what do you think about `ntp_leap_indicator`? I prefer it.

darkk commented:
Is there anything that is not addressed so far?

discordianfish commented on collector/ntp.go:
```diff
@@ -17,62 +17,145 @@ package collector
+			nil, nil,
+		), prometheus.GaugeValue},
+		reftime: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "ntp", "reftime"),
```
This should have _seconds suffix too. Maybe call it reference_timestamp_seconds? All our timestamps are usually called xx_timestamp_seconds.

darkk replied:
> xx_timestamp_seconds

Okay. I just followed naming convention used by series I've seen
- stat_linux with `node_boot_time`
- time with `node_time`
There are also some series I've not seen
- exec_bsd with `...boot_timestamp_seconds`
- runint with `...state_last_change_timestamp_seconds`

I don't feel like there is consistency here, but I'm missing something for sure :) Also, it's a bit confusing that node_exporter exports `node_boot_time` and `...boot_timestamp_seconds` depending on kernel, but that's probably due to backward compatibility...

<troll-mode>
Should I rename `node_time` to `node_time_timestamp_seconds` as well? :-D
</troll-mode>

discordianfish commented:
Okay, I'm fine with keeping leap flag like it is. But feel like we should rename 'reftime' to 'reference_timestamp_seconds' to stay consistent. Beside that, LGTM!

darkk commented:
Should I update doc as a separate PR after #664 is merged or should these two PRs be merged in that order: merge #664, update docs/TIME.md, merge #655 ?

discordianfish commented:
@darkk Let's just see which one we get merged first. LGTM but need another one giving ok before we can merge.

SuperQ commented on collector/ntp.go:
```diff
@@ -17,62 +17,145 @@ package collector
-	ntpServer          = kingpin.Flag("collector.ntp.server", "NTP server to use for ntp collector.").Default("").String()
+	ntpServer          = kingpin.Flag("collector.ntp.server", "NTP server to use for ntp collector").Default("127.0.0.1").String()
 	ntpProtocolVersion = kingpin.Flag("collector.ntp.protocol-version", "NTP protocol version").Default("4").Int()
+	ntpServerIsLocal   = kingpin.Flag("collector.ntp.server-is-local", "Certify that collector.ntp.server address is the same local host as this collector.").Default("false").Bool()
```
Should this be `true` by default?

darkk replied:
Nope, default IP passes `IsLoopback` check. That's safety check to disarm if you want to poll something like 172.16.0.1.

SuperQ commented on collector/ntp.go:
```diff
@@ -17,62 +17,145 @@ package collector
+			nil, nil,
+		), prometheus.GaugeValue},
+		leap: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "ntp", "leap"),
```
Please declare the subsystem `"ntp"` as a const.

SuperQ commented:
Looking pretty good, a couple minor nits.

SuperQ commented:
@matthiasr or @grobie, would either of you be able to give this one last look?

discordianfish commented:
I guess they won't :) So lets merge this to get some feedback before the next release.
